File: src/python/tensors_load.py
import proto.tensors_pb2 as protoTensors
import numpy as np

# Load the serialized tensor from file
with open("tensor.bin", "rb") as f:
    serialized_tensor = f.read()

# Parse the serialized tensor into a NumTensor protobuf object
tensor_3d = protoTensors.NumTensor()
tensor_3d.ParseFromString(serialized_tensor)

# Unpack the NumTensor protobuf object into a 3-D numpy array
my_array = tensor_3d.my_arrays[0]  # get first my_array
Z, Y, X = len(tensor_3d.my_arrays), len(my_array.my_sub_arrays), len(my_array.my_sub_arrays[0].my_array)
tensor = np.zeros((Z, Y, X), dtype=np.int32)

for i in range(Z):
    my_array = tensor_3d.my_arrays[i]
    for j in range(Y):
        my_sub_array = my_array.my_sub_arrays[j]
        for k in range(X):
            tensor[i,j,k] = my_sub_array.my_array[k]

print(tensor)

File: src/cpp/tensors_load.cpp
#include <iostream>
#include <fstream>
#include <string>
#include "proto/tensors.pb.h"

int main() {
  // declare a numeric tensor from `NumTensor` protobuf
  proto::NumTensor tensor_3d;


  // deserialize the tensor from the file
  std::ifstream input("tensor.bin", std::ios::in | std::ios::binary);
  if (!tensor_3d.ParseFromIstream(&input)) {
    std::cerr << "Failed to read tensor." << std::endl;
    return -1;
  }

  // convert the tensor to a 3D numpy array
  int Z = tensor_3d.my_arrays_size();
  int Y = tensor_3d.my_arrays(0).my_sub_arrays_size();
  int X = tensor_3d.my_arrays(0).my_sub_arrays(0).my_array_size();
  std::cout << "Tensor shape: (" << Z << ", " << Y << ", " << X << ")" << std::endl;
  int data[Z][Y][X];
  for (int i = 0; i < Z; i++) {
    for (int j = 0; j < Y; j++) {
      for (int k = 0; k < X; k++) {
        data[i][j][k] = tensor_3d.my_arrays(i).my_sub_arrays(j).my_array(k);
      }
    }
  }

  // print the tensor
  for (int i = 0; i < Z; i++) {
    std::cout << "z = " << i << std::endl;
    for (int j = 0; j < Y; j++) {
      for (int k = 0; k < X; k++) {
        std::cout << data[i][j][k] << " ";
      }
      std::cout << std::endl;
    }
  }

  return 0;
}